src/prototype_rust_sdk/workflow_context.rs

```diff
@@ -0,0 +1,130 @@
+    /// commands, and a receiver which outputs commands sent from the workflow.
+    pub(super) fn new(
+        args: Vec<Payload>,
+        am_cancelled: watch::Receiver<bool>,
+    ) -> (Self, Receiver<RustWfCmd>) {
+        // We need to use a normal std channel since our receiving side is non-async
+        let (chan, rx) = crossbeam::channel::unbounded();
+        (
+            Self {
+                chan,
+                args,
+                am_cancelled,
+            },
+            rx,
+        )
+    }
+
+    /// Get the arguments provided to the workflow upon execution start
+    pub fn get_args(&self) -> &[Payload] {
+        self.args.as_slice()
+    }
+
+    /// A future that resolves if/when the workflow is cancelled
+    pub async fn cancelled(&mut self) {
+        if *self.am_cancelled.borrow() {
+            return;
+        }
+        self.am_cancelled
+            .changed()
+            .await
+            .expect("Cancelled send half not dropped")
+    }
+
+    /// Request to create a timer
+    pub fn timer(&mut self, a: StartTimer) -> impl Future<Output = ()> {
+        let (cmd, unblocker) = WFCommandFut::new();
+        self.send(
+            CommandCreateRequest {
+                cmd: a.into(),
+                unblocker,
+            }
+            .into(),
+        );
+        cmd.map(|ue| {
+            if let UnblockEvent::Timer(_) = ue {
+            } else {
+                panic!("Wrong unblock event")
+            }
+        })
+    }
+
+    /// Request to run an activity
+    pub fn activity(&mut self, a: ScheduleActivity) -> impl Future<Output = ActivityResult> {
+        let (cmd, unblocker) = WFCommandFut::new();
+        self.send(
+            CommandCreateRequest {
+                cmd: a.into(),
+                unblocker,
+            }
+            .into(),
+        );
+        cmd.map(|ue| {
+            if let UnblockEvent::Activity { result, .. } = ue {
+                result
+            } else {
+                panic!("Wrong unblock event")
+            }
+        })
+    }
+
+    /// Cancel a timer
+    pub fn cancel_timer(&self, timer_id: &str) {
+        self.send(RustWfCmd::CancelTimer(timer_id.to_string()))
+    }
+
+    /// Cancel activity
+    pub fn cancel_activity(&self, activity_id: &str) {
+        self.send(RustWfCmd::CancelActivity(activity_id.to_string()))
+    }
+
+    /// Force a workflow task timeout by waiting too long and gumming up the entire runtime
+    pub fn force_timeout(&self, by_waiting_for: Duration) {
+        self.send(RustWfCmd::ForceTimeout(by_waiting_for))
+    }
+
+    fn send(&self, c: RustWfCmd) {
+        self.chan.send(c).unwrap();
+    }
+}
+
+struct WFCommandFut {
+    result_rx: oneshot::Receiver<UnblockEvent>,
+}
+
+impl WFCommandFut {
+    fn new() -> (Self, oneshot::Sender<UnblockEvent>) {
+        let (tx, rx) = oneshot::channel();
+        (Self { result_rx: rx }, tx)
```

Comment:
Isn't this a bit confusing that `new` returns something other than `Self`?

Reply:
Not necessarily. Pretty common pattern for things that have "handles" associated with them